### PR TITLE
Topic recognition - Jarrod Mann COMP3710 s4696612

### PR DESCRIPTION
**Pull request by jarrodmann**
Hi,
This pull request contains a brain MRI super-resolution network implementation. The model, which was implemented in Pytorch, uses a number of convolutional layers followed by a pixel shuffle layer. A single epoch of training was completed on the ADNI brain dataset to upscale images by a factor of 4, resulting in mean squared error loss of approximately 0.1 in the test set. Further details regarding training and results are available in the README file.

The pull request contains the following files:
READEME.md : documentation of the project.
modules.py : the code for the model used for super-resolution.
train.py : the code to train, validate, save and then test a model.
predict.py : the code to visualise the effectiveness of a model created by train.py.
utils.py : additional utilities to support other code files.
dataset.py : the code for loading the data images and rescaling them for training.
training_loss.png : a graph showing loss throughout training. For use in README.md.
model_sample.png : a sample of model created images. For use in README.md.
goal_images.png : a sample of goal images. For use in README.md.
Comparison.png : images comparing model results with goal images. For use in README.md.

Thanks,
Jarrod Mann - 46966122

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

Difficulty: Normal

**General Comments:** PixelCNN is implemented in PyTorch, along with dataloaders and a training script. While functional, the model is not changed substantially. Dataloaders require the user to manually create train and test folders, otherwise ok. Output images have different contrast to target. Script also only runs through a single epoch and doesn't correctly track validation loss. This is reflected in the generally low-quality images in the README. No metrics are cited for the performance of the model, relying solely on the provided sample images and the (faulty) training loss.

**[Recognition Problem]**
"**modules.py**":

- Contains a relatively simple PyTorch implementation of PixelCNN that is similar to the Keras example provided, only notable changes are the very large number of filters (400), larger filter sizes, and inclusion of relu activation functions. 

"**dataset.py**":

- Dataloader expects images to be in a folder, rather than split by AD and NC. Will require user to extract images into separate train and test folders manually rather than using the provided file structure; expected file structure not mentioned in readme, but does mention where to change path locations. 
- Correctly resizes the image to 64, 60.

"**train.py**":

- Train and validation are generated randomly? 
- Uses one of the training dataset batches to generate preview images rather than testing dataset.
- Only keeps the worst loss from validation dataset, rather than tracking the loss.
- Runs through a whole epoch of the validation set for each batch of the training set.
- Only one epoch run, meaning metrics in readme are just tracking the batch-to-batch loss over a single epoch for training, and the worst loss in the validation set for each batch of the training dataset.
- No metrics tracked for the test dataset.
- Correctly saves model weights and creates figures.

"**predict.py**":

- Correctly loads stored network weights.
- File only generates images from the first batch, does not track any losses over the test set.

"**README.MD**":

- Provides a description of the network and some motivation to the problem.
- Dependencies listed.
- Example inputs and outputs are provided.

**[Commit Log]**
Commit log ok, seems to update the git repository following changes/improvements/development of the code.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
